syntax = "proto3";
package swiftsignals.simulation;

option go_package = "protos/gen/simulation";

service SimulationService {
  rpc GetSimulationResults(SimulationRequest)
      returns (SimulationResultsResponse);
  rpc GetSimulationOutput(SimulationRequest) returns (SimulationOutputResponse);
}

message SimulationRequest {
  string intersection_id = 1;
  SimulationParameters simulation_parameters = 2;
}

message SimulationParameters {
  IntersectionType intersection_type = 1;
  int32 green = 2;
  int32 yellow = 3;
  int32 red = 4;
  int32 speed = 5;
  int32 seed = 6;
}

enum IntersectionType {
  INTERSECTION_TYPE_UNSPECIFIED = 0;
  INTERSECTION_TYPE_TRAFFICLIGHT = 1;
  INTERSECTION_TYPE_TJUNCTION = 2;
  INTERSECTION_TYPE_ROUNDABOUT = 3;
  INTERSECTION_TYPE_STOP_SIGN = 4;
}

message SimulationResultsResponse {
  int64 total_vehicles = 1;
  float average_travel_time = 2;
  float total_travel_time = 3;
  float average_speed = 4;
  float average_waiting_time = 5;
  float total_waiting_time = 6;
  int64 generated_vehicles = 7;
  int64 emergency_brakes = 8;
  int64 emergency_stops = 9;
  int64 near_collisions = 10;
}

message SimulationOutputResponse {
  Intersection intersection = 1;
  repeated Vehicle vehicles = 2;
}

message Intersection {
  repeated Node nodes = 1;
  repeated Edge edges = 2;
  repeated Connection connections = 3;
  repeated TrafficLight traffic_lights = 4;
}

message Node {
  string id = 1;
  float x = 2;
  float y = 3;
  NodeType type = 4;
}

enum NodeType {
  PRIORITY = 0;
  TRAFFIC_LIGHT = 1;
}

message Edge {
  string id = 1;
  string from = 2;
  string to = 3;
  float speed = 4;
  int32 lanes = 5;
}

message Connection {
  string from = 1;
  string to = 2;
  int32 fromLane = 3;
  int32 toLane = 4;
  int32 tl = 5;
}

message TrafficLight {
  string id = 1;
  string type = 2;
  repeated Phase phases = 3;
}

message Phase {
  int32 duration = 1;
  string state = 2;
}

message Vehicle {
  string id = 1;
  repeated Position positions = 2;
}

message Position {
  int32 time = 1;
  float x = 2;
  float y = 3;
  float speed = 4;
}
